File: Unique Path 2/SpaceOptimization.java

public class SpaceOptimization {
    public static void main(String[] args) {
        int[][] arr = {
            {0,0,0},
            {0,1,0},
            {0,0,0}
        };


        int m = arr.length;
        int n = arr[0].length;

        int[] dp = new int[n];

        for(int i = 0 ; i < m ; i ++) {

            int[] temp =  new int[n];

            for(int j = 0; j < n;j++) {
                if(arr[i][j] == 1) continue;
                if(i == 0 &&j == 0) temp[0] = 1;
                else {
                    temp[j] = dp[j];
                    if(j > 0) temp[j] += temp[j-1];
                }
            }

            dp = temp;
        }

        System.out.println(dp[n-1]);

    }
}

File: Cherry Pickup II/SpaceOptimization.py
def cherryPikup(arr,m,n):

    dp = [  [0 for _ in range(n)] for _ in range(n) ] 
    
    for j1 in range(n):
        for j2 in range(n):
            if j1 == j2: dp[j1][j2] = arr[m-1][j2]
            else: dp[j1][j2] =  arr[m-1][j1] + arr[m-1][j2]

    for i in range(m-2,-1,-1):
        temp = [  [0 for _ in range(n)] for _ in range(n) ] 

        for j1 in range(n):
            for j2 in range(n):
                maxi = -1e8
                for dj1 in range(-1,2):
                    for dj2 in range(-1,2):
                        value = 0
                        if j1 ==  j2: value = arr[i][j1]
                        else: value = arr[i][j1] +  arr[i][j2]

                        if dj1 + j1 >= 0 and dj1 + j1 < n and dj2 + j2 >= 0 and dj2 + j2 < n:
                            value += dp[dj1 + j1][dj2 + j2]

                        maxi = max(maxi, value)
                
                temp[j1][j2] = maxi
        dp = temp

    return dp[0][n-1]

if __name__ == "__main__":
    arr = [
        [3, 1, 1],
        [2,5,1],
        [1, 5, 5],
        [2, 1, 1]
    ]

    m = len(arr)
    n = len(arr[0])

    print(cherryPikup(arr,m,n))